File: rust/project/ports.rb
module RustProjection
  module Projector
    module_function

    # `PortOperation` has no `references`/`givens`/`mutations`/`ensures`
    # at all (its own header: "a port is the anti-corruption boundary that
    # turns an external call into a fact... not a second place business
    # rules live") — so `command_skip_reason`'s own long list of ungenerable
    # SHAPES mostly does not apply here; the one real one that still can:
    # an attribute type this domain's own aggregate-level check never
    # resolved a Rust type for (a port introducing a value object no OTHER
    # attribute anywhere in this aggregate already forced into existence —
    # no real operation in this corpus does that; flagged rather than
    # silently assumed impossible). A list attribute carrying `admits:`/
    # `pattern:` used to be refused here too (`constraint_list_problems`) —
    # removed (docs/decisions/0051): confirmed against Ruby's real
    # dispatch pipeline that neither is ever enforced on a list attribute
    # regardless, so refusing to generate the operation at all was MORE
    # restrictive than Ruby, not a real gap to guard.
    #
    # NO LONGER REQUIRES A `reference_to <owner>` ATTRIBUTE — routing
    # separation (`to:`/`with:`) supplies the receiver identity externally
    # now, the same way an acting COMMAND's own identity does
    # (`registry.rb`'s `CommandInvocation#split_aggregate_receiver`), so a
    # self-reference on the operation is a migration-era spelling of its
    # receiver at most, never a requirement. `Pizzas::Order.PaymentGateway.
    # Receive` — the corpus's one live port operation — declares no such
    # reference at all and used to be skipped here as a codegen bug; it
    # generates for real now.
    def port_operation_skip_reason(operation, _owner_name, value_objects_by_name)
      unresolved = operation[:attributes].reject do |attr|
        next true if attr[:list] # a list attribute's element type is checked below
        next true if reference_type?(attr[:type])
        next true if SCALAR.key?(attr[:type])

        value_objects_by_name.key?(attr[:type])
      end
      return "attribute type(s) #{unresolved.map { |a| a[:type] }.uniq.join(', ')} not generated yet " \
             "(a value object this aggregate's own attributes never resolved a Rust type for)" if unresolved.any?

      nil
    end

    # ── ONE emitter per port operation — the args struct plus a pure
    # dispatch function with no repo, no Hydrate, no mutation record: an
    # operation neither hydrates nor saves an aggregate instance
    # (`PortOperationInterpreter#call`'s own shorter `DISPATCH_ORDER`,
    # ported the same way `emit_entity_command` ports `EntityInterpreter
    # #call`'s). `registry.rb`'s own `emit_reference_check` runs BEFORE
    # this function is even called (the same "needs `store`, only exists
    # at the router level" reason a command's own reference checks do,
    # `reactions.rb`'s header) — by the time control reaches here, the
    # referenced aggregate is already known to exist, so this only builds
    # the event(s) the operation's own `emits` names, addressed by
    # `receiver_id` — the router's own resolved receiver identity
    # (`CommandInvocation#split_aggregate_receiver`, registry.rb), passed
    # in as this function's own first parameter now rather than read off a
    # declared attribute. The payload this function's own events carry is
    # a placeholder (`Json::Null`) — `registry.rb`'s generated call site
    # wraps the return with the SAME `stamp_payload(events, &payload)` a
    # command's own dispatch call already does, which REPLACES it with the
    # router's real, raw facts — so this function never needs to build a
    # correct payload itself, only correct `name`/`aggregate`/`id`.
    #
    # A SELF-REFERENCE ATTRIBUTE (`reference_to <owner>`, the migration-era
    # spelling of this same receiver) is EXCLUDED from the generated args
    # struct/JSON codec/event payload entirely — it names no external fact
    # once routing supplies the receiver, the same reason `registry.rb`'s
    # `emit_registry` strips it from `facts_json` via `split_aggregate_
    # receiver`'s own `legacy_receiver_field` when one is present.
    def emit_port_operation(operation, port_name, owner_name, domain_name, value_objects_by_name, aggregates_by_name)
      args_struct = "#{rust_ident(port_name)}#{rust_ident(operation[:name])}Args"
      qualified   = "#{domain_name}::#{owner_name}"
      fact_attrs  = operation[:attributes].reject { |attr| reference_target(attr[:type]) == owner_name }

      struct_lines = ["pub struct #{args_struct} {"]
      fact_attrs.each do |attr|
        type = rust_type(attr[:type], list: attr[:list])
        type = "Option<#{type}>" if attr[:optional]
        struct_lines << "    #{Exemplar.render('struct_field', 'TmplFieldType' => type, 'tmpl_field' => rust_ident_field(attr[:name]))}"
      end
      struct_lines << "}"

      invariant_checks = invariant_checks_for(operation, aggregates_by_name, value_objects_by_name)
      fn = "#{port_name.downcase}_#{dispatch_fn_name(rust_ident(operation[:name]))}"

      events = operation[:emits].map do |event_name|
        "        crate::kernel::Event { name: #{event_name.inspect}.to_string(), aggregate: #{qualified.inspect}.to_string(), " \
          "id: receiver_id.to_string(), payload: crate::kernel::Json::Null, occurred_at: None, correlation: None },"
      end

      dispatch_fn = <<~RUST.rstrip
        pub fn dispatch_operation_#{fn}(receiver_id: &str, args: #{args_struct}) -> Result<Vec<crate::kernel::Event>, crate::kernel::Refusal> {
        #{invariant_checks.join("\n")}
            Ok(vec![
        #{events.join("\n")}
            ])
        }
      RUST

      [
        "#[derive(Debug, Clone)]\n#{struct_lines.join("\n")}",
        emit_to_json_flat(args_struct, fact_attrs, value_objects_by_name),
        emit_from_json_flat(args_struct, fact_attrs, value_objects_by_name, command_name: "#{port_name}.#{operation[:name]}",
                            interleave_checks: true, aggregates_by_name: aggregates_by_name),
        dispatch_fn,
      ].join("\n\n")
    end
  end
end

File: rust/codegen/src/ports.rs
//! Port of `rust/project/ports.rb` — read that file's own header comments
//! in full; this mirrors its algorithm directly, function for function.

use crate::exemplar::Exemplar;
use crate::json::Json;
use crate::naming;
use std::collections::HashMap;

/// A list attribute carrying `admits:`/`pattern:` used to be refused
/// here too (`constraint_list_problems`) — removed (docs/decisions/0051):
/// confirmed against Ruby's real dispatch pipeline that neither is ever
/// enforced on a list attribute regardless, so refusing to generate the
/// operation at all was MORE restrictive than Ruby, not a real gap.
pub fn port_operation_skip_reason(
    operation: &Json,
    _owner_name: &str,
    value_objects_by_name: &HashMap<String, &Json>,
) -> Option<String> {
    let attrs = operation.get("attributes").map(Json::each).unwrap_or(&[]);
    let unresolved: Vec<&Json> = attrs
        .iter()
        .filter(|attr| {
            if crate::attr::list(attr) {
                return false;
            }
            if naming::reference_type(crate::attr::type_name(attr)) {
                return false;
            }
            if naming::scalar_rust_type(crate::attr::type_name(attr)).is_some() {
                return false;
            }
            !value_objects_by_name.contains_key(crate::attr::type_name(attr))
        })
        .collect();
    if !unresolved.is_empty() {
        let mut types: Vec<&str> = Vec::new();
        for a in &unresolved {
            let t = crate::attr::type_name(a);
            if !types.contains(&t) {
                types.push(t);
            }
        }
        return Some(format!("attribute type(s) {} not generated yet (a value object this aggregate's own attributes never resolved a Rust type for)", types.join(", ")));
    }

    None
}

/// ── ONE emitter per port operation.
pub fn emit_port_operation(
    exemplar: &Exemplar,
    operation: &Json,
    port_name: &str,
    owner_name: &str,
    domain_name: &str,
    value_objects_by_name: &HashMap<String, &Json>,
    aggregates_by_name: &HashMap<String, &Json>,
) -> String {
    let args_struct = format!(
        "{}{}Args",
        naming::rust_ident(port_name),
        naming::rust_ident(operation.get("name").and_then(Json::as_str).unwrap_or(""))
    );
    let qualified = format!("{domain_name}::{owner_name}");
    let attrs = operation.get("attributes").map(Json::each).unwrap_or(&[]);
    // A self-reference on a port operation is a migration-era spelling of
    // its receiver, never an external fact. Keep accepting old IR while
    // excluding that field from the typed args and event payload.
    let fact_attrs: Vec<Json> = attrs
        .iter()
        .filter(|attr| naming::reference_target(crate::attr::type_name(attr)) != Some(owner_name))
        .cloned()
        .collect();

    let mut struct_lines = vec![format!("pub struct {args_struct} {{")];
    for attr in &fact_attrs {
        let mut ty = naming::rust_type(crate::attr::type_name(attr), crate::attr::list(attr));
        if crate::attr::optional(attr) {
            ty = format!("Option<{ty}>");
        }
        struct_lines.push(format!(
            "    {}",
            exemplar.render(
                "struct_field",
                &[
                    ("TmplFieldType", ty),
                    (
                        "tmpl_field",
                        naming::rust_ident_field(crate::attr::name(attr))
                    )
                ]
            )
        ));
    }
    struct_lines.push("}".to_string());

    let invariant_checks = crate::commands::invariant_checks_for(
        exemplar,
        operation,
        aggregates_by_name,
        value_objects_by_name,
    );
    let fn_name = format!(
        "{}_{}",
        port_name.to_lowercase(),
        naming::dispatch_fn_name(&naming::rust_ident(
            operation.get("name").and_then(Json::as_str).unwrap_or("")
        ))
    );

    let emits = operation.get("emits").map(Json::each).unwrap_or(&[]);
    let events: Vec<String> = emits
        .iter()
        .map(|event_name| {
            format!(
                "        crate::kernel::Event {{ name: {}.to_string(), aggregate: {}.to_string(), id: receiver_id.to_string(), payload: crate::kernel::Json::Null, occurred_at: None, correlation: None }},",
                naming::ruby_inspect_string(&event_name.to_s()),
                naming::ruby_inspect_string(&qualified)
            )
        })
        .collect();

    let dispatch_fn = format!(
        "pub fn dispatch_operation_{fn_name}(receiver_id: &str, args: {args_struct}) -> Result<Vec<crate::kernel::Event>, crate::kernel::Refusal> {{\n{}\n    Ok(vec![\n{}\n    ])\n}}",
        invariant_checks.join("\n"),
        events.join("\n")
    );

    let operation_name = operation.get("name").and_then(Json::as_str).unwrap_or("");
    [
        format!("#[derive(Debug, Clone)]\n{}", struct_lines.join("\n")),
        crate::json_codec::emit_to_json_flat(exemplar, &args_struct, &fact_attrs, value_objects_by_name, false, &[], None),
        crate::json_codec::emit_from_json_flat(
            exemplar,
            &args_struct,
            &fact_attrs,
            value_objects_by_name,
            None,
            Some(&format!("{port_name}.{operation_name}")),
            false,
            true,
            Some(aggregates_by_name),
        ),
        dispatch_fn,
    ]
    .join("\n\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn owner_receiver_is_not_a_port_fact_or_event_payload_field() {
        let operation = Json::parse(
            r#"{
              "name":"Receive",
              "attributes":[
                {"name":"order","type":"Reference<Order>","list":false,"optional":false},
                {"name":"amount","type":"Integer","list":false,"optional":false}
              ],
              "emits":["PaymentReceived"]
            }"#,
        )
        .unwrap();
        let value_objects = HashMap::new();
        let aggregates = HashMap::new();

        assert_eq!(
            port_operation_skip_reason(&operation, "Order", &value_objects),
            None
        );
        let generated = emit_port_operation(
            &Exemplar::load(),
            &operation,
            "PaymentGateway",
            "Order",
            "Pizzas",
            &value_objects,
            &aggregates,
        );

        assert!(generated.contains(
            "dispatch_operation_paymentgateway_receive(receiver_id: &str, args: PaymentGatewayReceiveArgs)"
        ));
        assert!(generated.contains("id: receiver_id.to_string()"));
        assert!(generated.contains("pub amount: i64"));
        assert!(!generated.contains("pub order:"));
        assert!(!generated.contains("v.require(\"order\""));
    }

    #[test]
    fn owner_operation_without_a_transitional_reference_is_generated() {
        let operation = Json::parse(
            r#"{
              "name":"Receive",
              "attributes":[{"name":"amount","type":"Integer","list":false,"optional":false}],
              "emits":["PaymentReceived"]
            }"#,
        )
        .unwrap();
        assert_eq!(
            port_operation_skip_reason(&operation, "Order", &HashMap::new()),
            None
        );
    }
}
